src/tangents.jl

```diff
@@ -436,16 +440,89 @@ end
     tangent_field_zeros_exprs = ntuple(fieldcount(P)) do n
         if tangent_field_type(P, n) <: PossiblyUninitTangent
             V = PossiblyUninitTangent{tangent_type(fieldtype(P, n))}
-            return :(isdefined(x, $n) ? $V(zero_tangent(getfield(x, $n))) : $V())
+            return :(isdefined(x, $n) ? $V(zero_tangent_internal(getfield(x, $n))) : $V())
         else
-            return :(zero_tangent(getfield(x, $n)))
+            return :(zero_tangent_internal(getfield(x, $n)))
         end
     end
     backing_data_expr = Expr(:call, :tuple, tangent_field_zeros_exprs...)
     backing_expr = :($(backing_type(P))($backing_data_expr))
     return :($(tangent_type(P))($backing_expr))
 end
 
+@inline zero_tangent_internal(x::Union{Int8,Int16,Int32,Int64,Int128,IEEEFloat}, ::IdDict) = zero_tangent_internal(x)
+@inline function zero_tangent_internal(x::SimpleVector, stackdict::IdDict)
+    return map!(n -> zero_tangent_internal(x[n], stackdict), Vector{Any}(undef, length(x)), eachindex(x))
+end
+@inline function zero_tangent_internal(x::Array{P, N}, stackdict::IdDict) where {P, N}
+    if haskey(stackdict, x)
+        return stackdict[x]
+    end
+    if isbitstype(P)
+        return _map_if_assigned!(Base.Fix2(zero_tangent_internal, stackdict), Array{tangent_type(P), N}(undef, size(x)...), x)
+    else
+        stackdict[x] = Array{tangent_type(P), N}(undef, size(x)...)
+        for i in eachindex(x)
+            if isassigned(x, i)
+                stackdict[x][i] = zero_tangent_internal(x[i], stackdict)
+            end
+        end
+        return stackdict[x]
+    end
+end
+function zero_tangent_internal(x::P, stackdict::IdDict) where {P<:Union{Tuple, NamedTuple}}
+    if haskey(stackdict, x)
+        return stackdict[x]
+    end
+    if tangent_type(P) == NoTangent
+        return NoTangent()
+    end 
+    stackdict[x] = tuple_map(Base.Fix2(zero_tangent_internal, stackdict), x)
+    return stackdict[x]
+end
+function zero_tangent_internal(x::P, stackdict::IdDict) where {P}
+    if haskey(stackdict, x)
+        return stackdict[x]
+    end
+
+    tangent_type(P) == NoTangent && return NoTangent()
+
+    if tangent_type(P) <: MutableTangent
+        stackdict[x] = tangent_type(P)()
+        zt = ntuple(fieldcount(P)) do n
+            name = fieldname(P, n)
+            if tangent_field_type(P, n) <: PossiblyUninitTangent
+                V = PossiblyUninitTangent{tangent_type(fieldtype(P, n))}
+                if isdefined(x, n)
+                    return (name => V(zero_tangent_internal(getfield(x, n), stackdict)))
+                else
+                    return (name => V())
+                end
+            else
+                return (name => zero_tangent_internal(getfield(x, n), stackdict))
+            end
+        end
+
+        stackdict[x].fields = NamedTuple(zt)
+        return stackdict[x]
+    else
+        zt = ntuple(fieldcount(P)) do n
```

Comment:
Should regular immutable `struct`s be put into the `IdDict`? I don't _think_ that they can ever alias one another / contain circular references, but would be happy to be shown wrong on this point!